src/session.rs

```diff
@@ -0,0 +1,342 @@
+            }
+        }
+        Ok(())
+    }
+
+    pub fn save_auth(&mut self, request_headers: &HeaderMap) -> Result<()> {
+        if let Some(value) = request_headers.get(AUTHORIZATION) {
+            self.content
+                .headers
+                .insert("authorization".into(), value.to_str()?.into());
+        }
+        Ok(())
+    }
+
+    pub fn cookies(&self) -> Vec<cookie_crate::Cookie> {
+        let mut cookies = vec![];
+        for (name, c) in self.content.cookies.iter() {
+            let mut cookie_builder = cookie_crate::Cookie::build(name, &c.value);
+            if let Some(expires) = c.expires {
+                cookie_builder =
+                    cookie_builder.expires(time::OffsetDateTime::from_unix_timestamp(expires));
+            }
+            if let Some(ref path) = c.path {
+                cookie_builder = cookie_builder.path(path);
+            }
+            if let Some(secure) = c.secure {
+                cookie_builder = cookie_builder.secure(secure);
+            }
+            cookies.push(cookie_builder.finish());
+        }
+        cookies
+    }
+
+    pub fn save_cookies(&mut self, cookies: Vec<cookie_crate::Cookie>) {
+        self.content.cookies.clear();
+        for cookie in cookies {
+            self.content.cookies.insert(
+                cookie.name().into(),
+                Cookie {
+                    value: cookie.value().into(),
+                    expires: cookie
+                        .expires()
+                        .and_then(|v| v.datetime())
+                        .map(|v| v.unix_timestamp()),
+                    path: cookie.path().map(Into::into),
+                    secure: cookie.secure(),
+                },
+            );
+        }
+    }
+
+    pub fn persist(&self) -> Result<()> {
+        if !self.path.exists() || !self.read_only {
+            if let Some(parent_path) = self.path.parent() {
+                fs::create_dir_all(parent_path)?;
+            }
+            let mut session_file = fs::File::create(&self.path)?;
+            let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
+            let mut ser = serde_json::Serializer::with_formatter(&mut session_file, formatter);
+            self.content.serialize(&mut ser)?;
+            session_file.write_all(b"\n")?;
+        }
+        Ok(())
+    }
+}
+
+fn is_path(value: &OsString) -> bool {
+    value.to_string_lossy().contains(std::path::is_separator)
+}
+
+fn config_dir() -> Option<PathBuf> {
+    if test_mode() {
+        Some(std::env::temp_dir())
+    } else {
+        dirs::config_dir()
+    }
+}
+
+fn xh_version() -> String {
+    if test_mode() {
+        "0.0.0".into()
+    } else {
+        env!("CARGO_PKG_VERSION").into()
+    }
+}
+
+fn path_from_url(url: &Url) -> Result<String> {
+    match (url.host_str(), url.port()) {
+        (Some(host), Some(port)) => Ok(format!("{}_{}", host, port)),
+        (Some(host), None) => Ok(host.into()),
```

Comment:
This is *technically* not quite safe because host might be `.` or `..` and then you're looking in the wrong directory.

But it seems very hard to run into, and HTTPie actually has the same issue:
```
$ echo ok > ~/.config/httpie/sessions/foo.json
$ http --session foo http://.
http: error: ConfigFileError: invalid session file: Expecting value: line 1 column 1 (char 0) [~/.config/httpie/sessions/foo.json]
```
So I'm also ok with ignoring it.

Reply:
This is now being handled by an early error